File: day01/ruby/day1_lib.rb
def calc_fuel(mass)
  ((mass.to_f / 3.0) - 2.0).floor
end


def recursive_fuel(mass)
  fuel_sum = 0
  last_mass = mass

  loop do
    fuel = [0,calc_fuel(last_mass)].max
    fuel_sum += fuel
    last_mass = fuel

    break if fuel == 0
  end

  fuel_sum
end

File: day1/c++/day1_lib.cpp
#include<math.h>
#include<algorithm>

int calc_fuel(int mass){
	return floor((mass / 3.0) - 2.0);
}

int recursive_fuel(int mass){
	int fuel_sum = 0;
	int last_mass = mass;
	int fuel = 0;

	do {
		fuel = std::max(0, calc_fuel(last_mass));
		fuel_sum += fuel;
		last_mass = fuel;
	} while (fuel != 0);

	return fuel_sum;
}
